Requestify generic signatures

CodaFi opened the pull request:
This just does the thing that @slavapestov's patch made trivial.  The real work is refining that assertion to read

```cxx
assert(!GenericSigAndBit.getInt() && "Generic signature cannot be changed");
```

so we can catch where we're double-computing signatures (*cough* validateDeclForNameLookup).

CodaFi commented:
@swift-ci please smoke test and merge

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
Jenkins crashed...

@swift-ci please smoke test macOS platform

slavapestov commented:
This is a great first step, but we also want to refactor the various places under validateDecl() where we call setGenericSignature() -- we ought to build the signature inside the evaluate() method.

slavapestov commented on include/swift/AST/NameLookupRequests.h:
```diff
@@ -273,6 +274,27 @@ class GenericParamListRequest :
   void cacheResult(GenericParamList *value) const;
 };
 
+class GenericSignatureRequest :
```
This should be in TypeCheckRequests since the implementation of evaluate() will use the functions in TypeCheckGeneric.cpp.

CodaFi replied:
The generic signature is computed by the verifier so clients not linking Sema all fail.

slavapestov commented on lib/AST/Decl.cpp:
```diff
@@ -856,18 +856,21 @@ GenericParamList *GenericContext::getGenericParams() const {
-    return GenericSig;
+    return GenericSignature::get({self}, {req});
   }
 
```
This should be moved to TypeCheckGeneric.cpp, and you can handle NominalTypeDecl, TypeAliasDecl, ExtensionDecl, AbstractFunctionDecl and SubscriptDecl there too by calling into the existing code.

slavapestov commented on include/swift/AST/NameLookupRequests.h:
```diff
@@ -27,6 +27,7 @@ class ClassDecl;
 class DestructorDecl;
 class GenericContext;
 class GenericParamList;
+class GenericSignature;
```
Still needed here?

slavapestov commented on lib/AST/GenericSignatureBuilder.cpp:
```diff
@@ -5296,6 +5296,9 @@ class GenericSignatureBuilder::InferRequirementsWalker : public TypeWalker {
     auto decl = ty->getAnyNominal();
     if (!decl) return Action::Continue;
 
+    if (!decl->hasComputedGenericSignature())
```
Why is this necessary? I would expect that kicking off a request from the GSB is totally the right thing to do

CodaFi replied:
This is a weird compromise.  The GSB does cycle detection, and so does the request evaluator.  We wind up double-diagnosing because of it in a few tests.

slavapestov replied:
We should clean this up later. Can you file a radar?

CodaFi replied:
rdar://55263708.  Will attach in a FIXME as well.

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -3781,15 +3802,15 @@ void TypeChecker::validateDecl(ValueDecl *D) {
     DeclValidationRAII IBV(typeAlias);
 
-    validateGenericTypeSignature(typeAlias);
+    (void)typeAlias->getGenericSignature();
```
This shouldn't be called from validateDecl(). You should force the request in typeCheckDecl() (or rather DeclChecker::visitTypeAliasDecl(), etc, instead)

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -4024,11 +4043,18 @@ void TypeChecker::validateDecl(ValueDecl *D) {
     
     // We want the function to be available for name lookup as soon
     // as it has a valid interface type.
+    (void)FD->getGenericSignature();
```
Again, let's not force the request here

slavapestov commented on lib/Sema/TypeCheckDecl.cpp:
```diff
@@ -4559,10 +4449,8 @@ void TypeChecker::validateExtension(ExtensionDecl *ext) {
-      auto *sig = checkExtensionGenericParams(*this, ext, genericParams);
-      ext->setGenericSignature(sig);
-    }
+    // Force the generic signature to be computed.
```
I think it should now be possible to remove validateExtension() altogether

slavapestov commented on lib/Sema/TypeCheckProtocolInference.cpp:
```diff
@@ -200,7 +200,7 @@ AssociatedTypeInference::inferTypeWitnessesViaValueWitnesses(
     // recursion breaker, in which case we can't yet confidently reject its
     // witnesses.
-    if (!extension->getGenericSignature())
+    if (!extension->hasComputedGenericSignature())
```
What happens if you just remove this?

CodaFi replied:
Asking for the generic signature kicks off type resolution which kicks off protocol witness  resolution which kicks off protocol inference which kicks off this request again. 😢

slavapestov replied:
Ah, something to untangle later then. Can you file a radar?

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -649,32 +608,119 @@ void TypeChecker::validateGenericTypeSignature(GenericTypeDecl *typeDecl) {
+    if (auto subscr = dyn_cast<SubscriptDecl>(accessor->getStorage())) {
+      return subscr->getGenericSignature();
+    }
+    llvm_unreachable("Wat?");
```
Convert the above dyn_cast to a cast, then remove this unreachable

slavapestov commented on lib/Sema/TypeCheckGeneric.cpp:
```diff
@@ -649,32 +608,119 @@ void TypeChecker::validateGenericTypeSignature(GenericTypeDecl *typeDecl) {
+    // Form the interface type of the extension so we can use it as an inference
+    // source.
+    //
+    // FIXME: Push this into the "get interface type" request.
```
Actually I'd like to redo how the inference works here so that formExtensionInterfaceType() is not needed at all.

slavapestov commented:
Really excited to see this come together!

CodaFi commented:
Tried to remove `validateExtension` last night and it led to a bunch of fallout and other cleanup that clutters this PR massively. Going to address your feedback then see if I can  use this to unblock @xymus’ PR because validateDeclForNameLookup stands in our way.

CodaFi committed:
Requestify generic signatures

Define a request that provides the generic signature for a given generic context.  This unblocks a ton of cleanup and will allow us to remove validateExtension, validateDeclForNameLookup, and a lot of the surrounding infrastructure.

Being more honest about which declarations actually have a generic signature computed has naturally introduced more cycles in requests.  hasComputedGenericSignature() now acts as a recursion breaker.  In the future, we should purge any uses of this accessor that specifically head-off cycles as the cycle itself is probably part of a larger structural problem.

CodaFi committed:
Teach the ASTVerifier not to force generic signatures

The ASTVerifier is called by parsing tests and IDE tests that may not necessarily set up a typechecker.  Remove the verifier's ability to force computing generic signatures so we don't accidentally wander into the typechecker and crash.

CodaFi committed:
Break some cycles

Computing the generic signature changes the way that cycles appear in the compiler.  For now, break these cycles.  We should investigate each place where hasComputedGenericSignature() is used in service of breaking cycles.  See rdar://55263708

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test

CodaFi commented:
@swift-ci please test source compatibility

CodaFi commented:
@swift-ci please test source compatibility

CodaFi commented:
Passed Swift, need LLDB patches.

CodaFi commented:
@swift-ci please test source compatibility

CodaFi commented:
@swift-ci please test source compatibility

CodaFi commented:
@swift-ci please smoke test

CodaFi commented:
Alright, at least locally this is sufficient hedging to get things to build

Canary first

@swift-ci please smoke test

CodaFi commented:
Cool, back to LLDB failures.  Off to compat

@swift-ci please test source compatibility

CodaFi commented:
apple/swift-lldb#1991
@swift-ci please test

CodaFi commented:
⛵️
